Constituency search

jonleesy opened the pull request:
Closes #77 
Closes #95

jonleesy committed:
Large Commit. Working on windows and forgot that there is no shortcut commit setup.

jonleesy committed:
Update map.js

jonleesy commented:
Will clean up the code once it's all approved.

kymckay commented on public/src/map.js:
```diff
@@ -22,10 +22,27 @@ export async function initMap(apiKey) {
+  
+  const constituencyGeocoder = new MapboxGeocoder({
     accessToken: apiKey,
+    localGeocoder: dummy,
```
What is this localGeocoder?

jonleesy replied:
On my side, at least, `localGeocoder` seemed to be whatever thats built into the package and available locally, if `localGeocoder = localSearch`, then it will return this error 
![image](https://user-images.githubusercontent.com/65177718/112053774-a9f34580-8b4c-11eb-968b-9fa2d5514363.png)

Hence stumbled upon some solution online

kymckay commented on public/src/map.js:
```diff
@@ -35,7 +52,36 @@ export async function initMap(apiKey) {
+  
+  function localSearch(query) {
+    const matchingFeatures = [];
+    return fetch('https://opendata.arcgis.com/datasets/937997590f724a398ccc0100dbd9feee_0.geojson')
```
This seems super inefficient (fetching the entire GeoJSON file with every search) and as it's being fetched from a 3rd party service is a bit of an abuse of their resources

Smart browsers probably cache the data anyway and it isn't an issue, but it's not good to rely on that

jonleesy replied:
Not really sure how else to do it. Actually followed this tutorial [https://docs.mapbox.com/mapbox-gl-js/example/forward-geocode-custom-data/](url) and somehow still did not get things to work. I used a `const data = require('<link>')`.

Was googling for answers and somehow stumbled upon this..

kymckay replied:
You managed to research and get it working which is the important thing, but I might see if the link there makes any sense to me later on as I wouldn't feel comfortable merging with the request like this

kymckay replied:
Also, you can't use `require` in client-side code, it is a Node.js interface

jonleesy replied:
yep 👍 understood

kymckay commented on public/src/map.js:
```diff
@@ -35,7 +52,36 @@ export async function initMap(apiKey) {
+                matchingFeatures.push(feature);
+              }
+          });
+          return Promise.resolve(matchingFeatures);
```
I believe you can just return `matchingFeatures` directly

jonleesy replied:
done!

kymckay reviewed (changes requested):
Nice work with the promises and getting this working!

Functions pretty well, but I think the way it's being achieved is significantly inefficient to the point that it's not worth it for this feature.

However, it really feels like there should be a simpler way using the package features, but you probably know more than I do about the Mapbox GL JS package at this point.

Even if there isn't, we can still make this such that it doesn't need to fetch the data file for every search and that would be a significant improvement to start with.

jonleesy committed:
Update map.js

jonleesy committed:
Merge branch 'main' into constituency-search

jonleesy committed:
Restructure of code

jonleesy committed:
Simplifying code. Use ISO3166 instead of hard coded bounds

jonleesy committed:
Update map.js

jonleesy committed:
Restructure

jonleesy committed:
Update map.js

jonleesy committed:
Update map.js

jonleesy committed:
Should solve check fail – white spaces

jonleesy committed:
No idea what made the test fail. Removed the comma

jonleesy commented:
There is that tidied up. Tried a few things with the search but didn't really work

kymckay committed:
Merge branch 'main' into constituency-search

kymckay committed:
Simplify supplementary constituency search

- Use localGeocoder for supplementary results from the data
- Avoid loading the data a second time, just use the source data from
  the map

kymckay committed:
Merge branch 'main' into constituency-search

kymckay commented:
I simplified the constituency geocoding, Mapbox's documentation is really not the greatest so a lot of my understanding was from Github tickets and the like.

The main improvement is not having to fetch and store the GeoJSON, just using the data from the map. There's an outstanding bug with duplicate results, but I think we get this merged and then fix it separately.
